text/0000-dst-coercion.md

```diff
@@ -0,0 +1,184 @@
+
+DST is not really finished without this, in particular there is a need for types
+like reference counted trait objects (`Rc<Trait>`) which are not currently well-
+supported (without coercions, it is pretty much impossible to create such values
+with such a type).
+
+# Detailed design
+
+There is an `Unsize` trait and lang item. This trait signals that a type can be
+converted using the compiler's coercion machinery from a sized to an unsized
+type. All implementations of this trait are implicit and compiler generated. It
+is an error to implement this trait. If `&T` can be coerced to `&U` then there
+will be an implementation of `Unsize<U>` for `T`. E.g, `[i32; 42]:
+Unsize<[i32]>`. Note that the existence of an `Unsize` impl does not signify a
+coercion can itself can take place, it represents an internal part of the
+coercion mechanism (it corresponds with `coerce_inner` from  RFC 401). The trait
+is defined as:
+
+```
+#[lang="unsize"]
+trait Unsize<T: ?Sized>: ::std::marker::PhantomFn<Self, T> {}
+```
+
+There are implementations for any fixed size array to the corresponding unsized
+array, for any type to any trait that that type implements, for structs and
+tuples where the last field can be unsized, and for any pair of traits where
+`Self` is a sub-trait of `T` (see RFC 401 for more details).
+
+There is a `CoerceUnsized` trait which is implemented by smart pointer types to
+opt-in to DST coercions. It is defined as:
+
+```
+#[lang="coerce_unsized"]
+trait CoerceUnsized<Target>: ::std::marker::PhantomFn<Self, Target> + Sized {}
+```
+
+An example implementation:
+
+```
+impl<T: ?Sized+Unsize<U>, U: ?Sized> CoerceUnsized<Rc<U>> for Rc<T> {}
+impl<T: ?Sized+CoerceUnsized<U>, U: ?Sized> CoerceUnsized<NonZero<U>> for NonZero<T> {}
+
+// For reference, the definitions of Rc and NonZero:
+pub struct Rc<T: ?Sized> {
+    _ptr: NonZero<*mut RcBox<T>>,
+}
+pub struct NonZero<T: Zeroable+?Sized>(T);
+```
+
+Implementing `CoerceUnsized` indicates that the self type should be able to be
+coerced to the `Target` type. E.g., the above implementation means that
+`Rc<[i32; 42]>` can be coerced to `Rc<[i32]>`. There will be `CoerceUnsized` impls
+for the various pointer kinds available in Rust and which allow coercions, therefore
+`CoerceUnsized` when used as a bound indicates coercible types. E.g.,
+
+```
+fn foo<T: CoerceUnsized<U>, U>(x: T) -> U {
+    x
+}
+```
+
+Built-in pointer impls:
+
+```
+impl<T: ?Sized+Unsize<U>, U: ?Sized> CoerceUnsized<Box<U>> for Box<T> {}
+impl<T: ?Sized+Unsize<U>, U: ?Sized, 'a> CoerceUnsized<&'a U> for Box<T> {}
+impl<T: ?Sized+Unsize<U>, U: ?Sized, 'a> CoerceUnsized<&mut 'a U> for Box<T> {}
+impl<T: ?Sized+Unsize<U>, U: ?Sized> CoerceUnsized<*const U> for Box<T> {}
+impl<T: ?Sized+Unsize<U>, U: ?Sized> CoerceUnsized<*mut U> for Box<T> {}
+
+impl<T: ?Sized+Unsize<U>, U: ?Sized, 'a, 'b: 'a> CoerceUnsized<&'a U> for &mut 'b U {}
+impl<T: ?Sized+Unsize<U>, U: ?Sized, 'a> CoerceUnsized<&mut 'a U> for &mut 'a U {}
```

Comment:
I think this is true. I'll probably leave the RFC as is, since that reflects the implementation. But it does feel we could make this more flexible.

cc @nikomatsakis